src/crimson/osd/osd.cc

```diff
@@ -823,6 +830,100 @@ seastar::future<> OSD::handle_pg_log(ceph::net::Connection* conn,
   return do_peering_event(m->get_spg(), std::move(evt));
 }
 
+seastar::future<> OSD::handle_fast_pg_create(ceph::net::Connection* conn,
+                                             Ref<MOSDPGCreate2> m)
+{
+  if (!m->get_source().is_mon()) {
+    logger().warn("got pg create request from non-mon. ignoring");
+    return seastar::now();
+  }
+  return seastar::parallel_for_each(m->pgs,
+    [m,this](pair<spg_t,pair<epoch_t,utime_t>> p) {
+      auto& [pgid, pg_create] = p;
+
+      const auto q = m->pg_extra.find(pgid);
+      if (q == std::end(m->pg_extra)) {
+        // pre-octopus ... no pg history.  this can be removed in Q release.
+        // XXX: peering machinery bypassed. Handling create_info directly.
+        auto& [created_epoch, created_stamp] = pg_create;
+        pg_history_t fallback_pg_history{created_epoch, created_stamp};
+        return handle_pg_create_info(osdmap, std::make_unique<PGCreateInfo>(
+	  pgid, created_epoch, fallback_pg_history, PastIntervals{}, true));
+      }
+
+      const auto& [pg_history, past_intervals] = q->second;
+      if (!past_intervals.empty() &&
+	  m->epoch < past_intervals.get_bounds().second) {
+        logger().error("got pg_create on {} epoch {} and unmatched "
+                       "past_intervals {} (history {})",
+                       pgid, m->epoch, past_intervals, pg_history);
+      } if (const auto found = pgs.find(pgid); found != pgs.end()) {
+        logger().info("pg {} already exists. Ignoring creation request", pgid);
+      } else if (osdmap->is_up_acting_osd_shard(pgid, whoami)) {
+        logger().info("pg {} does not exist. Creating", pgid);
+        // XXX: peering machinery bypassed. Handling create_info directly.
+        return handle_pg_create_info(osdmap, std::make_unique<PGCreateInfo>(
+	  pgid, osdmap->get_epoch(), pg_history, past_intervals, true));
+      }
+      return seastar::now();
```

Comment:
nit, i'd suggest return `seastar::now()` explicitly in the error paths, instead of having it as a `catch-all` airbag.